language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/performance/CodeActionPerformanceTest.java

```diff
@@ -0,0 +1,196 @@
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package org.ballerinalang.langserver.performance;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import org.ballerinalang.langserver.codeaction.CodeActionUtil;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.commons.LanguageServerContext;
+import org.ballerinalang.langserver.commons.workspace.WorkspaceDocumentException;
+import org.ballerinalang.langserver.commons.workspace.WorkspaceManager;
+import org.ballerinalang.langserver.contexts.LanguageServerContextImpl;
+import org.ballerinalang.langserver.util.FileUtils;
+import org.ballerinalang.langserver.util.TestUtil;
+import org.ballerinalang.langserver.workspace.BallerinaWorkspaceManager;
+import org.eclipse.lsp4j.CodeActionContext;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.Endpoint;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Test performance of code actions feature in language server.
+ */
+public class CodeActionPerformanceTest {
+    private Endpoint serviceEndpoint;
+    private final JsonParser parser = new JsonParser();
+    private final Path sourcesPath
+            = new File(getClass().getClassLoader().getResource("performance").getFile()).toPath();
+    private static final WorkspaceManager workspaceManager
+            = BallerinaWorkspaceManager.getInstance(new LanguageServerContextImpl());
+    private static final LanguageServerContext serverContext = new LanguageServerContextImpl();
+
+    @BeforeClass
+    public void init() throws Exception {
+        this.serviceEndpoint = TestUtil.initializeLanguageSever();
+    }
+
+    @Test(dataProvider = "performance-data-provider")
+    public void testCodeAction(String config, String source) throws IOException, WorkspaceDocumentException {
```

Comment:
Can you please elaborate?

Reply:
What @nadeeshaan means here is regarding code duplication. Since we have similar methods/snippets in existing code action and completion tests, we should try to reuse them by moving common logic into a utility class like `TestUtil`.